server/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java

```diff
@@ -178,7 +178,7 @@ protected void doStart(ClusterState clusterState) {
                                 @Override
                                 public void handleException(final TransportException exp) {
                                     Throwable cause = exp.unwrapCause();
-                                    if (cause instanceof ConnectTransportException) {
+                                    if (cause instanceof ConnectTransportException || cause instanceof NodeClosedException) {
```

Comment:
Can you write an integration test that shows that this solves an issue?

Perhaps a test that restarts the current master while concurrently sending some `TransportMasterNodeReadAction` (e.g. cluster health).

I'm surprised that we have not seen this issue in any of our integration tests, where we sometimes restart nodes (but perhaps don't do this to concurrently issuing master-level requests).

Reply:
There we go: 3378d3de2a248e87e30c4f67315842322c90f432

Bit of a dirty test but reliably reproduces the issue without any other hacks :)

I've researched this a little and the problem seems to be isolated to `7.x+`. When shutting down a master there is a short window during which:
1. We will get the cluster state without master node from `org.elasticsearch.cluster.coordination.Coordinator#clusterStateWithNoMasterBlock`
2. That then causes the master action **on the current master** node to fail with `NotMasterException` and add the cluster state observer for the retry waiting for the new master. That new master never comes and instead the wait is shut-down with the `NodeClosedException` when the cluster service is shut down
3. Client receives `NodeClosedException` back from the master

=> this fix still seems fine, if we retry on `NodeClosedException` because we can interpret it as the master node going away we're good IMO.

Reply:
> the problem seems to be isolated to 7.x+

Does that also include 7.0, 7.1, ...?
I would like to understand why this is not an issue in earlier versions.

Reply:
For snapshots this only started showing up because I inadvertently disabled the snapshot shard's services own retry in https://github.com/elastic/elasticsearch/pull/50788 

I can also reproduce this failure back to `7.1`, though interestingly enough it seems to be a lot less likely in `7.1` than in  `7.6` (maybe that's due to some locking we removed from the internal test cluster operations, but I can't tell right now).

I pushed https://github.com/elastic/elasticsearch/pull/51325/commits/646823d17d0f1cd972d9dd2b6c316674ba23a572 to slow things down a little more and make the test repro better. Without this change it takes a pretty large number of iterations for it to fail on `7.1`